### ISSUE TITLE
tflite output is NaN

### ISSUE THREAD

**Issue by 1alexuwu** [tensorflow/tflite-support#844]
Hello,

When I get the float array output of my tflite model, I get NaN. What could be the problem? I am still new to do this so there is still a lot that I don't know. I am trying to use my tflite model to the application that I am making. I hope you can help me. 

Extracting keypoints from mediapipe:

      `private ArrayList<Float> getMultiHandLandmarksDebugString(List<NormalizedLandmarkList> multiHandLandmarks) {
              int numHands = multiHandLandmarks.size();
              ArrayList<Float> hands = new ArrayList<>();

            if (numHands == 1){
                ArrayList<Float> anotherHand = new ArrayList<>();
                for(NormalizedLandmarkList landmarks : multiHandLandmarks){
                    for(NormalizedLandmark landmark : landmarks.getLandmarkList()){
                        hands.add((Float) landmark.getX());
                        hands.add((Float) landmark.getY());
                        hands.add((Float) landmark.getZ());
                    }
                }

                for (int j = 0; j < 21; j++){
                    anotherHand.add((float) 0);
                    anotherHand.add((float) 0);
                    anotherHand.add((float) 0);
                }

                hands.addAll(anotherHand);
            }
            else if (numHands == 2){
                for(NormalizedLandmarkList landmarks: multiHandLandmarks){
                    for(NormalizedLandmark landmark: landmarks.getLandmarkList()){
                        hands.add((Float) landmark.getX());
                        hands.add((Float) landmark.getY());
                        hands.add((Float) landmark.getZ());
                    }
                }
            }

        return hands;
    }`

All keypoints holder:

`private ArrayList<ArrayList<Float>> allKeypoints = new ArrayList<>();`

Code for getting keypoints from camera and prediction:

    `processor.addPacketCallback(
                    OUTPUT_LANDMARKS_STREAM_NAME,
                    (packet) -> {
                            List<NormalizedLandmarkList> multiHandLandmarks =
                                    PacketGetter.getProtoVector(packet, NormalizedLandmarkList.parser());

                        allKeypoints.add(getMultiHandLandmarksDebugString(multiHandLandmarks));

                        if (allKeypoints.size() == 30){
                            System.out.println(allKeypoints.size());
                            System.out.println(allKeypoints);

                            //convert to 1d array
                            float[] keypointsArray = new float[allKeypoints.size() * 126];
                            int i = 0;
                            for (ArrayList<Float> a: allKeypoints){
                                for (Float b: a){
                                    keypointsArray[i] = b;
                                    i++;
                                }
                            }

                            allKeypoints.clear();

                            try {
                                Alphabet30Frames model = Alphabet30Frames.newInstance(getApplicationContext());

                                // Creates inputs for reference.
                                TensorBuffer inputFeature0 = TensorBuffer.createFixedSize(new int[]{1, 30, 126}, DataType.FLOAT32);
                                ByteBuffer byteBuffer = ByteBuffer.allocateDirect(30*126*4);
                                byteBuffer.asFloatBuffer().put(keypointsArray);
                                inputFeature0.loadBuffer(byteBuffer);

                                // Runs model inference and gets result.
                                Alphabet30Frames.Outputs outputs = model.process(inputFeature0);
                                TensorBuffer outputFeature0 = outputs.getOutputFeature0AsTensorBuffer();

                                float[] confidences = outputFeature0.getFloatArray();

                                for (int v = 0; v < confidences.length; v++){
                                    System.out.println(confidences[v]);
                                }

                                // Releases model resources if no longer used.
                                model.close();
                            } catch (IOException e) {
                                // TODO Handle the exception
                            }
                        }else {
                            System.out.println("Collecting");
                        }
                });`

**Comment by lu-wang-g** [comment]
If you don't get any error message, at least your model runs. Have you verified your model can run correctly, i.e. run it using the [tflite Python interpreter](https://www.tensorflow.org/lite/guide/inference#load_and_run_a_model_in_python)?

**Comment by 1alexuwu** [comment]
I tested the tflite model using python interpreter and it works just fine. I found a solution here https://github.com/tensorflow/tensorflow/issues/22803?fbclid=IwAR20GVY0stOBiqdaWVtJK0d9H7J80NbCRK7iAnt9gcO6BDJIxX46QDxfVKo . The tflite model is quantized (if my term is right?). It works fine now in the android app.

**Comment by himikewu** [comment]
> I tested the tflite model using python interpreter and it works just fine. I found a solution here [tensorflow/tensorflow#22803](https://github.com/tensorflow/tensorflow/issues/22803) . The tflite model is quantized (if my term is right?). It works fine now in the android app.

can you describe you solution more clear?

**Comment by alex-pascual** [comment]
I used this code to convert my model to TFlite.

from tensorflow import keras
import tensorflow as tf
model = keras.models.load_model(r"path to file")
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS]
tflite_model = converter.convert()
open("file.tflite", "wb").write(tflite_model)

I found it here https://github.com/tensorflow/tensorflow/issues/22803.

I think it has something to do with quantizing or optimizing the TFlite for devices with low processing power.

**Comment by himikewu** [comment]
> I used this code to convert my model to TFlite.
> 
> from tensorflow import keras import tensorflow as tf model = keras.models.load_model(r"path to file") converter = tf.lite.TFLiteConverter.from_keras_model(model) converter.optimizations = [tf.lite.Optimize.DEFAULT] converter.target_spec.supported_ops = [tf.lite.OpsSet.TFLITE_BUILTINS] tflite_model = converter.convert() open("file.tflite", "wb").write(tflite_model)
> 
> I found it here [tensorflow/tensorflow#22803](https://github.com/tensorflow/tensorflow/issues/22803).
> 
> I think it has something to do with quantizing or optimizing the TFlite for devices with low processing power.

Yes ,I found the quantized tflite model can work well on computer by , but  have some problems on arduino. I use the lastest arduino tensorflow library . I guess the TFLM's operators can't be compatible with quantized tflite model.

**Comment by jakobkilian** [comment]
Struggled with this issue, too. So just wanted to leave a note:
If you use [Dynamic Range Quantization](https://www.tensorflow.org/lite/performance/post_training_quantization#dynamic_range_quantization) in the conversion **without** further quantization step you will run in to problem! Idk if this is intended behaviour. So here are my observations:

Using Dynamic Range Quantization I could only run super tiny models with like 2 hidden layers and 32 float nodes and 1 in and out on an ESP32:
```
converter = tf.lite.TFLiteConverter.from_keras_model(model)
converter.optimizations = [tf.lite.Optimize.DEFAULT]
tflite_model = converter.convert()
```

Using **either no optimizations** like so
```
converter = tf.lite.TFLiteConverter.from_keras_model(model)
tflite_model = converter.convert()
```
**or doing the full quantization thing** (with `converter.representative_dataset` and so on) I could run hundreds of nodes without a problem.
